### ISSUE TITLE
False positive `redundant-expr`, but only for `if x and y:`, not for `if x: if y:`

### ISSUE THREAD

**Issue by MarcoGorelli** [python/mypy#21533]
**Bug Report**

In the code below
```py
if not _is_seq_of(a, Cat) and not _is_seq_of(a, int):
```
is treated differently to
```py
if not _is_seq_of(a, Cat):
    if not _is_seq_of(a, int):
```

**To Reproduce**

https://mypy-play.net/?mypy=latest&python=3.14&enable-error-code=redundant-expr&gist=470e934d38b23a68087d2cd4ebcb9a2a

```python
from typing import TypeIs, TypeVar, Sequence, Any, Self, reveal_type, Iterable, TypeAlias

class Cat:
    def foo(self) -> Self:  # type: ignore[empty-body]
        ...


MyType = TypeVar('MyType', int, str, Cat)
T = TypeVar('T')

def _is_seq_of(seq: Sequence[Any], tp: type[T]) -> TypeIs[Sequence[T]]:  # type: ignore[empty-body]
    ...


def main1(a: Sequence[MyType], how: str) -> MyType:
    if how.startswith('align'):
        if not _is_seq_of(a, Cat) and not _is_seq_of(a, int):
            msg = 'unexpected'
            raise TypeError(msg)
    return a[0]

def main2(a: Sequence[MyType], how: str) -> MyType:
    if how.startswith('align'):
        if not _is_seq_of(a, Cat):
            if not _is_seq_of(a, int):
                msg = 'unexpected'
                raise TypeError(msg)
    return a[0]
```

Mypy accepts `main2` just fine, but for `main1`, it reports
```
$ mypy --enable-error-code redundant-expr t.py
t.py:17: error: Left operand of "and" is always true  [redundant-expr]
Found 1 error in 1 file (checked 1 source file)
```

**Expected Behavior**

I think `main1` and `main2` should be treated the same. My expectation is that they should both pass (not report any errors)

**Actual Behavior**

```console
$ mypy --enable-error-code redundant-expr t.py
t.py:17: error: Left operand of "and" is always true  [redundant-expr]
Found 1 error in 1 file (checked 1 source file)
```

**Your Environment**

- Mypy version used: 2.1.0
- Mypy command-line flags: --enable-error-code redundant-expr
- Mypy configuration options from `mypy.ini` (and other config files):
- Python version used: 3.14.4

**Comment by hauntsaninja** [comment]
I believe this is currently intentional.

If implemented in the most straight forward way, it would trigger on a lot of runtime checking (e.g. `if not isinstance(x, int): raise TypeError`). For it to be work well in practice, the check needs a similar implementation to unreachable code, where we forgive unreachable code that always raises. But this is a little trickier to implement, for reasons. (We could also just not solve that problem and use another more pedantic error code. Relevant code in `visit_if_stmt` if someone wants to play around)

**Comment by piyush-003** [comment]
I investigated this issue and traced the root cause to find_isinstance_check returning Never instead of the correct type for the false case of TypeIs with constrained TypeVars. Specifically in conditional_types function in checker.py, when current_type is already a subtype of proposed_type, it returns UninhabitedType() for the no case — which is incorrect for constrained TypeVars.

I attempted a workaround by setting right_always=True in visit_if_stmt when the body always raises, but this had unintended side effects on type narrowing in many real-world projects.

Would it be appropriate to fix the root cause in find_isinstance_check or conditional_types? Is there a safer way to handle this specific case with constrained TypeVars and TypeIs?
